### PR TITLE
Add e2e  benchmarks for head and block Queries

### PR DESCRIPTION
**Pull request by krasi-georgiev**
closes #240 
~closes #278~ already closed in https://github.com/prometheus/tsdb/pull/572

combined and refactored the above PRs.

will post results from the benchmarks shortly.

Signed-off-by: Krasi Georgiev <kgeorgie@redhat.com>

### CODE REVIEW

**Comment by krasi-georgiev** [comment]
```
benchcmp no-expand.txt expand.txt
benchmark                                                                 old ns/op     new ns/op     delta
BenchmarkPersistedQueries_Series_1M_RESelector_1_Expansion-8              157304440     147421099     -6.28%
BenchmarkPersistedQueries_Series_1M_RESelector_2_Expansion-8              19163203      17982639      -6.16%
BenchmarkPersistedQueries_Series_1M_RESelector_3_Expansion-8              8640811       7914559       -8.40%
BenchmarkPersistedQueries_3Blocks_Series_1M_RESelector_1_Expansion-8      515304845     471832010     -8.44%
BenchmarkPersistedQueries_3Blocks_Series_1M_RESelector_2_Expansion-8      64759526      69743052      +7.70%
BenchmarkPersistedQueries_3Blocks_Series_1M_RESelector_3_Expansion-8      29525021      28743209      -2.65%
BenchmarkPersistedQueries_10Blocks_Series_1M_RESelector_1_Expansion-8     572205794     580150933     +1.39%
BenchmarkPersistedQueries_10Blocks_Series_1M_RESelector_2_Expansion-8     72159910      82287972      +14.04%
BenchmarkPersistedQueries_10Blocks_Series_1M_RESelector_3_Expansion-8     33388668      26196775      -21.54%

benchmark                                                                 old allocs     new allocs     delta
BenchmarkPersistedQueries_Series_1M_RESelector_1_Expansion-8              1000182        1000223        +0.00%
BenchmarkPersistedQueries_Series_1M_RESelector_2_Expansion-8              110304         110325         +0.02%
BenchmarkPersistedQueries_Series_1M_RESelector_3_Expansion-8              40215          40232          +0.04%
BenchmarkPersistedQueries_3Blocks_Series_1M_RESelector_1_Expansion-8      3100695        3100832        +0.00%
BenchmarkPersistedQueries_3Blocks_Series_1M_RESelector_2_Expansion-8      341929         341997         +0.02%
BenchmarkPersistedQueries_3Blocks_Series_1M_RESelector_3_Expansion-8      124624         124676         +0.04%
BenchmarkPersistedQueries_10Blocks_Series_1M_RESelector_1_Expansion-8     3100696        3100830        +0.00%
BenchmarkPersistedQueries_10Blocks_Series_1M_RESelector_2_Expansion-8     341929         341996         +0.02%
BenchmarkPersistedQueries_10Blocks_Series_1M_RESelector_3_Expansion-8     124624         124676         +0.04%

benchmark                                                                 old bytes     new bytes     delta
BenchmarkPersistedQueries_Series_1M_RESelector_1_Expansion-8              55219483      59879260      +8.44%
BenchmarkPersistedQueries_Series_1M_RESelector_2_Expansion-8              6081791       6468332       +6.36%
BenchmarkPersistedQueries_Series_1M_RESelector_3_Expansion-8              2208200       2324264       +5.26%
BenchmarkPersistedQueries_3Blocks_Series_1M_RESelector_1_Expansion-8      144936442     158922606     +9.65%
BenchmarkPersistedQueries_3Blocks_Series_1M_RESelector_2_Expansion-8      15965682      17126937      +7.27%
BenchmarkPersistedQueries_3Blocks_Series_1M_RESelector_3_Expansion-8      5801820       6150190       +6.00%
BenchmarkPersistedQueries_10Blocks_Series_1M_RESelector_1_Expansion-8     144937071     158921465     +9.65%
BenchmarkPersistedQueries_10Blocks_Series_1M_RESelector_2_Expansion-8     15965733      17126544      +7.27%
BenchmarkPersistedQueries_10Blocks_Series_1M_RESelector_3_Expansion-8     5801675       6150330       +6.01%
```

**Comment by krasi-georgiev** [comment]
ping @codesome 	@gouthamve 	@brian-brazil 	
ready for review (thanks to @VladMasarik for fixing a small bug)

**Comment by brian-brazil** [comment]
Hmm, I've some benchmarks in #572 also. We should de-duplicate a bit I think.

File: querier_test.go [hunk 1]
```diff
@@ -1467,6 +1467,322 @@ func (m mockIndex) LabelNames() ([]string, error) {
...
+			}
+			testutil.Ok(b, ss.Err())
+		}
+	}
+}
+
+var series100000 = genSeries(100000, 2, 0, 10, false)
+var series1000Random = genSeries(1000, 2, 0, 20, true)
+
+func BenchmarkInMemQueries_Series_EQSelector_1_Expansion(b *testing.B) {
```
Comment (brian-brazil) [inline querier_test.go]: This would be clearer with sub-benchmarks.

**Comment by krasi-georgiev** [comment]
some extensive query benchmarks

cc @gouthamve @codesome

**Comment by krasi-georgiev** [comment]
@brian-brazil couldn't have a quick look and let me know if you still think we should merge this with `BenchmarkHeadPostingForMatchers`

**Comment by brian-brazil** [comment]
As I said on IRC there, it'd be best to run the postings benchmarks both ways.

**Comment by krasi-georgiev** [comment]
What do you mean by `both ways`?

**Comment by brian-brazil** [comment]
With just the head, and with blocks.

**Comment by krasi-georgiev** [comment]
These benchmarks also test head and blocks completely separately.
https://github.com/prometheus/tsdb/pull/467/files#diff-a1431a302431f51c7f9e7c0b3cb58ed0R2197

I also don't mind to keep these separately.

**Comment by brian-brazil** [comment]
Ah right, I'd missed that.

**Comment by brian-brazil** [comment]
:+1:

**Pull request merged**